File: Sources/Application/TrafficMonitorState.swift
import Foundation

struct TrafficMonitorState: Equatable, Sendable {
  var connectionState: MonitorConnectionState = .disconnected
  var rates: CategorizedTrafficRates = .zero
  var rawRates: CategorizedTrafficRates = .zero
  var coverage: Double?
  var attributionCoverage = ConnectionAttributionCoverage.empty
  var liveProxyConnections: [LiveTrafficConnection] = []
  var liveDirectConnections: [LiveTrafficConnection] = []
  var activeProxyLeaves: [String] = []
  var activeRuleTypes: [String] = []
  var runtimeConfiguration: MihomoRuntimeConfiguration?
  var mihomoVersion: String?
  var lastObservedAt: Date?
  var message = "首次使用请先准备并连接 Mihomo；不知道它是什么也没关系。"

  mutating func resetLiveData() {
    rates = .zero
    rawRates = .zero
    coverage = nil
    attributionCoverage = .empty
    liveProxyConnections = []
    liveDirectConnections = []
    activeProxyLeaves = []
    activeRuleTypes = []
    lastObservedAt = nil
  }
}

struct TrafficStatusItemSnapshot: Equatable, Sendable {
  let rate: TrafficRate
  let connectionState: MonitorConnectionState
}

File: platform/windows/MihomoMeter.Windows.Core/Application/TrafficMonitorState.cs
using MihomoMeter.Windows.Core.Domain;

namespace MihomoMeter.Windows.Core.Application;

public enum MonitorConnectionState
{
    Disconnected,
    Connecting,
    Connected,
    Stale,
    Reconnecting,
    AuthenticationFailed,
    Unsupported,
}

public sealed record MihomoRuntimeConfiguration(
    string? Mode,
    bool? IsTunEnabled,
    string? TunStack,
    bool? AutomaticallyRoutesTraffic,
    bool? IsIPv6Enabled,
    bool? AllowsLan,
    int? MixedPort,
    MihomoProcessMatchingMode? ProcessMatchingMode);

public sealed record TrafficMonitorSnapshot(
    MonitorConnectionState State,
    string Message,
    string? MihomoVersion = null,
    CategorizedTrafficRates? Rates = null,
    double? Coverage = null,
    long SessionGeneration = 0,
    ConnectionAttributionCoverage AttributionCoverage = default,
    IReadOnlyList<LiveTrafficConnection>? ProxyConnections = null,
    IReadOnlyList<LiveTrafficConnection>? DirectConnections = null,
    MihomoProcessMatchingMode? ProcessMatchingMode = null,
    IReadOnlyList<string>? ProxyLeaves = null,
    IReadOnlyList<string>? RuleTypes = null,
    MihomoRuntimeConfiguration? RuntimeConfiguration = null)
{
    public IReadOnlyList<LiveTrafficConnection> LiveProxyConnections =>
        ProxyConnections ?? Array.Empty<LiveTrafficConnection>();

    public IReadOnlyList<LiveTrafficConnection> LiveDirectConnections =>
        DirectConnections ?? Array.Empty<LiveTrafficConnection>();

    public IReadOnlyList<string> ActiveProxyLeaves =>
        ProxyLeaves ?? Array.Empty<string>();

    public IReadOnlyList<string> ActiveRuleTypes =>
        RuleTypes ?? Array.Empty<string>();

    public static TrafficMonitorSnapshot Disconnected => new(
        MonitorConnectionState.Disconnected,
        "尚未连接 Mihomo Controller。");
}

public sealed record MonitoringPolicy(
    TimeSpan StaleAfter,
    TimeSpan ReconnectAfter,
    TimeSpan BackoffResetAfter)
{
    public static MonitoringPolicy Production => new(
        TimeSpan.FromSeconds(2),
        TimeSpan.FromSeconds(5),
        TimeSpan.FromSeconds(30));

    public MonitoringPolicy Validate()
    {
        if (StaleAfter <= TimeSpan.Zero
            || ReconnectAfter <= StaleAfter
            || BackoffResetAfter <= TimeSpan.Zero)
        {
            throw new ArgumentOutOfRangeException(nameof(MonitoringPolicy));
        }

        return this;
    }
}
